monai/transforms/nvtx.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Wrapper around NVIDIA Tools Extension for profiling MONAI transformations
+"""
+
+from monai.transforms.transform import RandomizableTransform, Transform
+from monai.utils import optional_import
+
+_nvtx, _ = optional_import("torch._C._nvtx", descriptor="NVTX is not installed. Are you sure you have a CUDA build?")
+
+__all__ = [
+    "Mark",
+    "Markd",
+    "MarkD",
+    "MarkDict",
+    "RandMark",
+    "RandMarkd",
+    "RandMarkD",
+    "RandMarkDict",
+    "RandRangePop",
+    "RandRangePopd",
+    "RandRangePopD",
+    "RandRangePopDict",
+    "RandRangePush",
+    "RandRangePushd",
+    "RandRangePushD",
+    "RandRangePushDict",
+    "RangePop",
+    "RangePopd",
+    "RangePopD",
+    "RangePopDict",
+    "RangePush",
+    "RangePushd",
+    "RangePushD",
+    "RangePushDict",
+]
+
+
+class RangePush(Transform):
+    """
+    Pushes a range onto a stack of nested range span.
+    Stores zero-based depth of the range that is started.
+
+    Args:
+        msg: ASCII message to associate with range
+    """
+
+    def __init__(self, msg: str) -> None:
+        self.msg = msg
+        self.depth = None
+
+    def __call__(self, data):
+        self.depth = _nvtx.rangePushA(self.msg)
+        return data
+
+
+class RandRangePush(RangePush, RandomizableTransform):
+    """
+    Pushes a range onto a stack of nested range span (RandomizableTransform).
+    Stores zero-based depth of the range that is started.
+
+    Args:
+        msg: ASCII message to associate with range
+    """
+
+
+class RangePop(Transform):
```

Comment:
according to the tutorial, I think those push/pop operations should be implemented as a python context manager for better readability, what do you think?

Reply:
@wyli, you're right but the reason that we have separate push/pop transforms is that the range that they defines is not limited to a transform and can include multiple transforms.

Reply:
I'll implement the context manager too.  #2754